### PR TITLE
internal/stubserver: Close Client Conn in error handling of Start

### PR DESCRIPTION
**Pull request by Mskxn**
This pr fixes a small blocking problem related to tests that use `stubserver.Start`. This function doesn't close the server/client when error ocurres, which causes some blocking routines.
It can be repro with leakcheck and  some delay to trigger the timeout in `waitForReady`

These tests below are influenced
```
--- FAIL: TestInvoke (26.84s)
--- FAIL: TestRetryTransparentWhenCommitted (26.88s)
--- FAIL: TestUnaryClientInterceptor_ContextValuePropagation (26.87s)
--- FAIL: TestDetailedConnectionCloseErrorPropagatesToRpcError (26.88s)
--- FAIL: TestNilStatsHandler (26.88s)
--- FAIL: TestClientCancellationPropagatesUnary (26.86s)
--- FAIL: TestInvokeCancelClosedNonFailFast (26.87s)
--- FAIL: TestServerReturningContextError (26.89s)
--- FAIL: TestChainStreamServerInterceptor (26.82s)
--- FAIL: TestStatusInvalidUTF8Message (26.86s)
--- FAIL: TestGzipBadChecksum (26.85s)
--- FAIL: TestUnarySetSendCompressorAfterHeaderSendFailure (26.87s)
--- FAIL: TestInvokeLargeErr (26.90s)
--- FAIL: TestGlobalBinaryLoggingOptions (26.87s)
--- FAIL: TestChainOnBaseUnaryServerInterceptor (26.85s)
--- FAIL: TestChainStreamClientInterceptor_ContextValuePropagation (26.86s)
--- FAIL: TestForceCodecName (26.87s)
--- FAIL: TestForceServerCodec (26.88s)
--- FAIL: TestGRPCMethod (26.93s)
--- FAIL: TestMetadataInAddressAttributes (26.89s)
``` 

one  of the leakcheck report, the others are similar
```
=== RUN   TestInvoke
    invoke_test.go:42: Failed to start stub server: context deadline exceeded
    leakcheck.go:115: Leaked goroutine: goroutine 10 [select]:
        google.golang.org/grpc.(*ccBalancerWrapper).watcher(0xc00006f740)
        	C:/Users/Msk/Documents/GitHub/grpc-go/balancer_conn_wrappers.go:115 +0x73
        created by google.golang.org/grpc.newCCBalancerWrapper
        	C:/Users/Msk/Documents/GitHub/grpc-go/balancer_conn_wrappers.go:76 +0x22a
    leakcheck.go:115: Leaked goroutine: goroutine 37 [IO wait]:
        internal/poll.runtime_pollWait(0x1fe6a8a2828, 0x72)
        	C:/Users/Msk/go/go1.20rc3/src/runtime/netpoll.go:306 +0x89
        internal/poll.(*pollDesc).wait(0x0?, 0x0?, 0x0)
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.execIO(0xc0000b4018, 0x14608c8)
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_windows.go:175 +0xf7
        internal/poll.(*FD).Read(0xc0000b4000, {0xc000216000, 0x8000, 0x8000})
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_windows.go:436 +0x2b8
        net.(*netFD).Read(0xc0000b4000, {0xc000216000?, 0x1040100000000?, 0x0?})
        	C:/Users/Msk/go/go1.20rc3/src/net/fd_posix.go:55 +0x29

        net.(*conn).Read(0xc00020a000, {0xc000216000?, 0x133ba00?, 0xc000235d68?})
        	C:/Users/Msk/go/go1.20rc3/src/net/net.go:183 +0x45
        bufio.(*Reader).Read(0xc0002000c0, {0xc0001da3c0, 0x9, 0x30?})
        	C:/Users/Msk/go/go1.20rc3/src/bufio/bufio.go:237 +0x1bb
        io.ReadAtLeast({0x150a240, 0xc0002000c0}, {0xc0001da3c0, 0x9, 0x9}, 0x9)
        	C:/Users/Msk/go/go1.20rc3/src/io/io.go:332 +0x9a
        io.ReadFull(...)
        	C:/Users/Msk/go/go1.20rc3/src/io/io.go:351
        golang.org/x/net/http2.readFrameHeader({0xc0001da3c0?, 0x9?, 0xc000286120?}, {0x150a240?, 0xc0002000c0?})
        	C:/Users/Msk/go/pkg/mod/golang.org/x/net@v0.8.0/http2/frame.go:237 +0x6e
        golang.org/x/net/http2.(*Framer).ReadFrame(0xc0001da380)
        	C:/Users/Msk/go/pkg/mod/golang.org/x/net@v0.8.0/http2/frame.go:498 +0x95
        google.golang.org/grpc/internal/transport.(*http2Client).reader(0xc00023a000, 0x0?)
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_client.go:1595 +0x257
        created by google.golang.org/grpc/internal/transport.newHTTP2Client
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_client.go:397 +0x1e0a
    leakcheck.go:115: Leaked goroutine: goroutine 38 [select]:
        google.golang.org/grpc/internal/transport.(*controlBuffer).get(0xc00020e050, 0x1)
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/controlbuf.go:418 +0x115
        google.golang.org/grpc/internal/transport.(*loopyWriter).run(0xc0001aa000)
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/controlbuf.go:552 +0x91
        google.golang.org/grpc/internal/transport.newHTTP2Client.func6()
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_client.go:451 +0x85
        created by google.golang.org/grpc/internal/transport.newHTTP2Client
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_client.go:449 +0x23dc
    leakcheck.go:115: Leaked goroutine: goroutine 54 [select]:
        google.golang.org/grpc/internal/transport.(*controlBuffer).get(0xc0000a8230, 0x1)
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/controlbuf.go:418 +0x115
        google.golang.org/grpc/internal/transport.(*loopyWriter).run(0xc000144070)
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/controlbuf.go:552 +0x91
        google.golang.org/grpc/internal/transport.NewServerTransport.func2()
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_server.go:341 +0xda
        created by google.golang.org/grpc/internal/transport.NewServerTransport
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_server.go:338 +0x1a73
    leakcheck.go:115: Leaked goroutine: goroutine 55 [select]:
        google.golang.org/grpc/internal/transport.(*http2Server).keepalive(0xc0000844e0)
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_server.go:1155 +0x233
        created by google.golang.org/grpc/internal/transport.NewServerTransport
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_server.go:344 +0x1ab8
    leakcheck.go:115: Leaked goroutine: goroutine 56 [IO wait]:
        internal/poll.runtime_pollWait(0x1fe6a8a2738, 0x72)
        	C:/Users/Msk/go/go1.20rc3/src/runtime/netpoll.go:306 +0x89
        internal/poll.(*pollDesc).wait(0x0?, 0x0?, 0x0)
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.execIO(0xc0000b4298, 0x14608c8)
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_windows.go:175 +0xf7
        internal/poll.(*FD).Read(0xc0000b4280, {0xc0000c8000, 0x8000, 0x8000})
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_windows.go:436 +0x2b8
        net.(*netFD).Read(0xc0000b4280, {0xc0000c8000?, 0x1040100000000?, 0x0?})
        	C:/Users/Msk/go/go1.20rc3/src/net/fd_posix.go:55 +0x29

        net.(*conn).Read(0xc0000ac008, {0xc0000c8000?, 0x0?, 0x0?})
        	C:/Users/Msk/go/go1.20rc3/src/net/net.go:183 +0x45
        bufio.(*Reader).Read(0xc00028e000, {0xc0000e0040, 0x9, 0x30?})
        	C:/Users/Msk/go/go1.20rc3/src/bufio/bufio.go:237 +0x1bb
        io.ReadAtLeast({0x150a240, 0xc00028e000}, {0xc0000e0040, 0x9, 0x9}, 0x9)
        	C:/Users/Msk/go/go1.20rc3/src/io/io.go:332 +0x9a
        io.ReadFull(...)
        	C:/Users/Msk/go/go1.20rc3/src/io/io.go:351
        golang.org/x/net/http2.readFrameHeader({0xc0000e0040?, 0x9?, 0xc0002860f0?}, {0x150a240?, 0xc00028e000?})
        	C:/Users/Msk/go/pkg/mod/golang.org/x/net@v0.8.0/http2/frame.go:237 +0x6e
        golang.org/x/net/http2.(*Framer).ReadFrame(0xc0000e0000)

        	C:/Users/Msk/go/pkg/mod/golang.org/x/net@v0.8.0/http2/frame.go:498 +0x95
        google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams(0xc0000844e0, 0x0?, 0x0?)
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/transport/http2_server.go:642 +0x167
        google.golang.org/grpc.(*Server).serveStreams(0xc0001123c0, {0x1514c78?, 0xc0000844e0})
        	C:/Users/Msk/Documents/GitHub/grpc-go/server.go:950 +0x189
        google.golang.org/grpc.(*Server).handleRawConn.func1()
        	C:/Users/Msk/Documents/GitHub/grpc-go/server.go:892 +0x46
        created by google.golang.org/grpc.(*Server).handleRawConn
        	C:/Users/Msk/Documents/GitHub/grpc-go/server.go:891 +0x185
    leakcheck.go:115: Leaked goroutine: goroutine 9 [IO wait]:
        internal/poll.runtime_pollWait(0x1fe6a8a2918, 0x72)
        	C:/Users/Msk/go/go1.20rc3/src/runtime/netpoll.go:306 +0x89
        internal/poll.(*pollDesc).wait(0x0?, 0x0?, 0x0)
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.execIO(0xc000114f18, 0xc000067c20)
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_windows.go:175 +0xf7
        internal/poll.(*FD).acceptOne(0xc000114f00, 0x2bc, {0xc0000a00f0?, 0xc000060c00?, 0x1461058?}, 0xc000067cd8?)
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_windows.go:936 +0x6d
        internal/poll.(*FD).Accept(0xc000114f00, 0xc000067df8)
        	C:/Users/Msk/go/go1.20rc3/src/internal/poll/fd_windows.go:970 +0x1d6
        net.(*netFD).accept(0xc000114f00)
        	C:/Users/Msk/go/go1.20rc3/src/net/fd_windows.go:139 +0x65
        net.(*TCPListener).accept(0xc0000089f0)
        	C:/Users/Msk/go/go1.20rc3/src/net/tcpsock_posix.go:148 +0x25
        net.(*TCPListener).Accept(0xc0000089f0)
        	C:/Users/Msk/go/go1.20rc3/src/net/tcpsock.go:297 +0x3d

        google.golang.org/grpc.(*Server).Serve(0xc0001123c0, {0x15103e0?, 0xc0000089f0})
        	C:/Users/Msk/Documents/GitHub/grpc-go/server.go:824 +0x475
        created by google.golang.org/grpc/internal/stubserver.(*StubServer).StartServer
        	C:/Users/Msk/Documents/GitHub/grpc-go/internal/stubserver/stubserver.go:115 +0x42e
--- FAIL: TestInvoke (10.06s)
```

RELEASE NOTES: none

### CODE REVIEW

File: internal/stubserver/stubserver.go [hunk 1]
```diff
@@ -125,13 +128,15 @@ func (ss *StubServer) StartClient(dopts ...grpc.DialOption) error {
 
 	cc, err := grpc.Dial(ss.Target, opts...)
 	if err != nil {
+		cc.Close()
```
Comment (easwars) [inline internal/stubserver/stubserver.go:131]: When `grpc.Dial()` returns a non-nil error, it is expected to return a `nil` ClientConn. So, this call to `Close()` in that case will `panic`.

Comment (Mskxn) [reply internal/stubserver/stubserver.go]: Right, fixed.

File: internal/stubserver/stubserver.go [hunk 2]
```diff
@@ -85,7 +85,10 @@ func (ss *StubServer) Start(sopts []grpc.ServerOption, dopts ...grpc.DialOption)
 	if err := ss.StartServer(sopts...); err != nil {
 		return err
 	}
-	return ss.StartClient(dopts...)
+	if err := ss.StartClient(dopts...); err != nil {
+		return err
+	}
+	return nil
```
Comment (easwars) [inline internal/stubserver/stubserver.go:91]: Why is this change required? The one-line previously was doing the same thing.

**Review by easwars — CHANGES_REQUESTED** [review]

File: internal/stubserver/stubserver.go (hunk 2 above)
Comment (Mskxn) [reply internal/stubserver/stubserver.go:91]: my fault, this addition is irrelevant here

**Review by easwars — APPROVED** [review]

**Comment by easwars** [comment]
@zasweq : For second set of eyes.

**Review by zasweq — APPROVED** [review]
LGTM.
